crates/iota-indexer/src/optimistic_indexing.rs

```diff
@@ -0,0 +1,393 @@
+// Copyright (c) 2025 IOTA Stiftung
+// SPDX-License-Identifier: Apache-2.0
+use std::{collections::BTreeMap, time::Duration};
+
+use diesel::{RunQueryDsl, sql_query, sql_types};
+use downcast::Any;
+use fastcrypto::{encoding::Base64, error::FastCryptoError, traits::ToFromBytes};
+use iota_json_rpc_types::{IotaTransactionBlockResponse, IotaTransactionBlockResponseOptions};
+use iota_rest_api::{ExecuteTransactionQueryParameters, client::TransactionExecutionResponse};
+use iota_types::{
+    base_types::TransactionDigest,
+    effects::TransactionEffectsAPI,
+    full_checkpoint_content::CheckpointTransaction,
+    signature::GenericSignature,
+    transaction::{Transaction, TransactionData},
+};
+
+use crate::{
+    errors::IndexerError,
+    handlers::{
+        TransactionObjectChangesToCommit,
+        checkpoint_handler::{CheckpointHandler, try_extract_df_kind},
+    },
+    indexer_reader::IndexerReader,
+    metrics::IndexerMetrics,
+    models::{
+        display::StoredDisplay,
+        event_indices::OptimisticEventIndices,
+        events::{OptimisticEvent, StoredEvent},
+        transactions::{OptimisticTransaction, StoredTransaction, TxGlobalOrder},
+        tx_indices::OptimisticTxIndices,
+    },
+    store::{IndexerStore, PgIndexerStore},
+    transactional_blocking_with_retry,
+    types::{
+        EventIndex, IndexedDeletedObject, IndexedEvent, IndexedObject, IndexedTransaction,
+        IndexerResult, IotaTransactionBlockResponseWithOptions, TxIndex,
+    },
+};
+
+type TransactionDataToCommit = (
+    OptimisticTransaction,
+    OptimisticTxIndices,
+    Vec<OptimisticEvent>,
+    OptimisticEventIndices,
+    BTreeMap<String, StoredDisplay>,
+    TransactionObjectChangesToCommit,
+);
+
+pub struct OptimisticTransactionExecutor {
+    fullnode_rest_client: iota_rest_api::Client,
+    indexer_reader: IndexerReader,
+    store: PgIndexerStore,
+    metrics: IndexerMetrics,
+}
+
+impl OptimisticTransactionExecutor {
+    pub fn new(
+        fullnode_rest_client: iota_rest_api::Client,
+        indexer_reader: IndexerReader,
+        store: PgIndexerStore,
+        metrics: IndexerMetrics,
+    ) -> Self {
+        Self {
+            fullnode_rest_client,
+            indexer_reader,
+            store,
+            metrics,
+        }
+    }
+
+    pub(crate) async fn execute_and_index_tx_effects(
```

Comment:
Nit: I think it is safe to name it
```suggestion
    pub(crate) async fn execute_and_index_transaction(
```
Indexing `tx_effects` is abstracted under `index_transaction`